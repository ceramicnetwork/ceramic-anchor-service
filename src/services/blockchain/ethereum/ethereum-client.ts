import {
  Address,
  ParseAbi, PublicClient,
} from "viem";

export interface ContractParameters {
  abi: ParseAbi<any>,
  functionName: string,
  account: Address,
  address: Address,
  args: any[],
  maxFeePerGas: bigint,
  maxPriorityFeePerGas: bigint,
}

export type BlockHash = Address

export type TransactionReceipt = {
  blockHash: BlockHash
  from: Address
  successful: boolean
}

export type Block = {
  blockHash: BlockHash | null
  blockNumber: bigint | null
  timestamp: bigint
}

export type FeeHistory = {
  baseFeePerGas: bigint
}

export interface EthereumClient {
  getChainId(): Promise<number>
  simulateContract(opts: ContractParameters): Promise<void>
  waitForTransactionReceipt(hash: Address): Promise<TransactionReceipt>
  getBlock(hash: Address): Promise<Block>
  getFeeHistory(): Promise<FeeHistory>
}

export class ViemEthereumClient implements EthereumClient {
  private readonly inner: PublicClient

  constructor(inner: PublicClient) {
    this.inner = inner
  }

  async getChainId(): Promise<number> {
    return await this.inner.getChainId()
  }

  async simulateContract(opts: ContractParameters): Promise<void> {
    await this.inner.simulateContract(opts)
  }

  async waitForTransactionReceipt(hash: Address): Promise<TransactionReceipt> {
    const res = await this.inner.waitForTransactionReceipt(({ hash: hash}))
    return {
      blockHash: res.blockHash,
      from: res.from,
      successful: res.status == 'success'
    }
  }

  async getBlock(hash: Address): Promise<Block> {
    const res = await this.inner.getBlock({blockHash: hash})
    return {
      blockHash: res.hash,
      blockNumber: res.number,
      timestamp: res.timestamp
    }
  }

  async getFeeHistory(): Promise<FeeHistory> {
    const res = await this.inner.getFeeHistory({
      blockCount: 4,
      rewardPercentiles: [25, 75]
    })
    if (res.baseFeePerGas.length == 0 || !res.baseFeePerGas[0]) {
      throw new Error('Unable to get baseFeePerGas')
    }
    return {
      baseFeePerGas: res.baseFeePerGas[0]
    }
  }
}
