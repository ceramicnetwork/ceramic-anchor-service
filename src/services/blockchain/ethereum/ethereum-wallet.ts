import {Address, WalletClient} from "viem";
import {ContractParameters} from "./ethereum-client.js";

export type TransactionHash = Address

export interface EthereumWallet {
  readonly address: Address
  writeContract(req: ContractParameters): Promise<TransactionHash>;
}

export class ViemEthereumWallet implements EthereumWallet {
  private readonly inner: WalletClient
  readonly address: Address

  constructor(inner: WalletClient, address: Address) {
    this.inner = inner
    this.address = address
  }

  public static async create(inner: WalletClient): Promise<EthereumWallet> {
    const walletAddress = (await inner.getAddresses()).at(0)
    if(!walletAddress) {
      throw new Error('No wallet addresses found')
    }
    return new ViemEthereumWallet(inner, walletAddress)
  }

  async writeContract(req: ContractParameters): Promise<TransactionHash> {
    return await this.inner.writeContract({
      abi: req.abi,
      functionName: req.functionName,
      address: req.address,
      account: req.account,
      chain: this.inner.chain,
    })
  }
}
