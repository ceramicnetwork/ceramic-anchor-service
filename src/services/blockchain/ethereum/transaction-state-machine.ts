import {ContractParameters, EthereumClient, TransactionReceipt} from "./ethereum-client.js";
import {Address, parseAbi, toHex} from "viem";
import {CID} from "multiformats/cid";
import {EthereumWallet} from "./ethereum-wallet.js";
import {backOff, BackoffOptions} from "exponential-backoff"
import { logger, logEvent } from '../../../logger/index.js'
import {Transaction} from "../../../models/transaction.js";

const FUNCTION_NAME = "anchorDagCbor" as const
const ABI = parseAbi([`function ${FUNCTION_NAME}(bytes32)`])

class PreviousAttempt {
  attemptNum: number
  maxFeePerGas: bigint
  maxPriorityFeePerGas: bigint

  constructor(baseFeePerGas: bigint) {
    this.attemptNum = 0

    this.maxPriorityFeePerGas = 1_500_000_000n // 1.5 gwei
    this.maxFeePerGas = this.calculateMaxFeePerGas(baseFeePerGas)
  }

  calculateMaxFeePerGas(baseFeePerGas: bigint): bigint {
    return (baseFeePerGas * 120n) / 100n + this.maxPriorityFeePerGas
  }

  increment(baseFeePerGas: bigint) {
    this.attemptNum += 1
    const increment = (100n + 10n * BigInt(this.attemptNum)) / 100n
    this.maxPriorityFeePerGas = this.maxPriorityFeePerGas * increment
    this.maxFeePerGas = this.calculateMaxFeePerGas(baseFeePerGas)
  }
}

interface GetFeeHistory {
  tag: 'fee_history'
}

interface SimulateContract {
  tag: 'simulate'
  opts: ContractParameters
}

interface WriteContract {
  tag: 'write'
  opts: ContractParameters
}

interface GetTransactionReceipt {
  tag: 'get_transaction'
  transaction: Address
}

interface GetBlock {
  tag: 'get_block'
  transaction: TransactionReceipt
}

const MAX_ATTEMPTS = 3

const BACKOFF_OPTIONS: BackoffOptions = {
  numOfAttempts: MAX_ATTEMPTS,
  delayFirstAttempt: false,
  startingDelay: 1_000, //delay in ms
  retry: (e, attemptNumber) => {
    logger.warn(`Failed to send transaction, attempt ${attemptNumber} of ${MAX_ATTEMPTS}: ${e}`)
    return true
  }
}

type TransactionState = GetFeeHistory | SimulateContract | WriteContract | GetTransactionReceipt | GetBlock
export class TransactionStateMachine {
  private readonly rootCid: CID
  private previousAttempt?: PreviousAttempt
  private state: TransactionState
  private provider: EthereumClient
  private wallet: EthereumWallet
  private readonly contractAddress: Address
  private readonly chainId: string

  constructor(chainId: string, provider: EthereumClient, wallet: EthereumWallet, contractAddress: Address, rootCid: CID) {
    this.chainId = chainId
    this.provider = provider
    this.wallet = wallet
    this.contractAddress = contractAddress
    this.rootCid = rootCid
    this.state = {
      tag: 'fee_history'
    }
  }

  async getFeeHistory(): Promise<void> {
    const fees = await backOff(() => this.provider.getFeeHistory(), BACKOFF_OPTIONS)
    const baseFeePerGas = fees.baseFeePerGas
    let previousAttempt = this.previousAttempt
    if (!previousAttempt) {
      previousAttempt = new PreviousAttempt(baseFeePerGas)
    } else {
      previousAttempt.increment(baseFeePerGas)
    }
    const data = toHex(this.rootCid.bytes.slice(4))
    const opts = {
      abi: ABI,
      functionName: FUNCTION_NAME,
      account: this.wallet.address,
      address: this.contractAddress,
      args: [data],
      maxFeePerGas: previousAttempt.maxFeePerGas,
      maxPriorityFeePerGas: previousAttempt.maxPriorityFeePerGas,
    }
    this.state = {
      tag: 'simulate' as const,
      opts
    }
  }

  async simulateContract(): Promise<void> {
    const state = this.state as SimulateContract
    await backOff(() => this.provider.simulateContract(state.opts), BACKOFF_OPTIONS)
    this.state = {
      tag: 'write',
      opts: state.opts
    }
  }

  async writeContract(): Promise<void> {
    const state = this.state as WriteContract
    const hash = await backOff(() => this.wallet.writeContract(state.opts), BACKOFF_OPTIONS)
    this.state = {
      tag: 'get_transaction',
      transaction: hash
    }
  }

  async getTransaction(): Promise<void> {
    const state = this.state as GetTransactionReceipt
    const tx = await backOff(() => this.provider.waitForTransactionReceipt(state.transaction), BACKOFF_OPTIONS)
    if (tx.successful) {
      this.state = {
        tag: 'get_block',
        transaction: tx,
      }
    } else {
      if (this.previousAttempt && this.previousAttempt.attemptNum < MAX_ATTEMPTS) {
        logger.warn(`Transaction failed, retrying (${this.previousAttempt.attemptNum} / ${MAX_ATTEMPTS})`)
      }
    }
  }

  async getBlock(): Promise<Transaction> {
    const state = this.state as GetBlock
    const block = await backOff(() => this.provider.getBlock(state.transaction.blockHash), BACKOFF_OPTIONS)
    if (!block.blockNumber) {
      throw new Error('Block did not have a block number')
    }

    logEvent.ethereum({
      type: 'txResponse',
      hash: state.transaction.blockHash,
      blockTimestamp: block.timestamp,
      blockNumber: block.blockNumber,
      blockHash: block.blockHash,
      from: state.transaction.from,
    })

    return new Transaction(this.chainId, state.transaction.blockHash, block.blockNumber, block.timestamp)

  }

  public async run(): Promise<Transaction> {
    for(;;) {
      if (this.state.tag == 'fee_history') {
        await this.getFeeHistory()
      } else if (this.state.tag == 'simulate') {
        await this.simulateContract()
      } else if (this.state.tag == 'write') {
        await this.writeContract()
      } else if (this.state.tag == 'get_transaction') {
        await this.getTransaction()
      } else if (this.state.tag == 'get_block') {
        return await this.getBlock()
      }
    }
  }
}
