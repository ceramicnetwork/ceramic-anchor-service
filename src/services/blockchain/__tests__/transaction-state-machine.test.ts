import { afterAll, beforeAll, beforeEach, describe, expect, jest, test } from '@jest/globals'
import {Block, ContractParameters, EthereumClient, FeeHistory, TransactionReceipt} from "../ethereum/ethereum-client.js"
import {Address} from "viem"
import {TransactionHash} from "../ethereum/ethereum-wallet.js"
import {TransactionStateMachine} from "../ethereum/transaction-state-machine.js";
import {CID} from "multiformats/cid";

const FROM = '0x1234'
const WALLET = '0x1234abcd'
const CONTRACT = '0xfeed'

class MockEthereumClient implements EthereumClient {
  async getChainId(): Promise<number> {
    return 1337
  }

  async simulateContract(opts: ContractParameters): Promise<void> {
    return
  }

  async waitForTransactionReceipt(hash: Address): Promise<TransactionReceipt> {
    return {
      blockHash: hash,
      from: FROM,
      successful: true,
    }
  }

  async getBlock(hash: Address): Promise<Block> {
    return {
      blockNumber: 1n,
      blockHash: hash,
      timestamp: 0n,
    }
  }

  async getFeeHistory(): Promise<FeeHistory> {
    return {
      baseFeePerGas: 100n,
    }
  }
}

class MockEthereumWallet {
  readonly address = WALLET
  async writeContract(req: ContractParameters): Promise<TransactionHash> {
    return '0xdeadbeef'
  }
}

describe('TransactionStateMachine with mocks', () => {
  test('no errors encountered', async () => {
    const provider = new MockEthereumClient()
    const wallet = new MockEthereumWallet()
    const cid = CID.parse('bafyreic5p7grucmzx363ayxgoywb6d4qf5zjxgbqjixpkokbf5jtmdj5ni')
    const tsm = new TransactionStateMachine('test', provider, wallet, CONTRACT, cid)
    const res = await tsm.run()
    expect(res).toMatchSnapshot()
  })

  test('simulate fails due to timeout', async () => {
    const provider = new MockEthereumClient()
    const wallet = new MockEthereumWallet()
    const simulateSpy = jest.spyOn(provider, 'simulateContract')
      .mockImplementationOnce(() => {
        throw new Error('Timeout')
      })
    const writeSpy = jest.spyOn(wallet, 'writeContract')
    const cid = CID.parse('bafyreic5p7grucmzx363ayxgoywb6d4qf5zjxgbqjixpkokbf5jtmdj5ni')
    const tsm = new TransactionStateMachine('test', provider, wallet, CONTRACT, cid)
    const res = await tsm.run()
    expect(res).toMatchSnapshot()
    expect(simulateSpy.mock.calls.length).toEqual(2)
    expect(writeSpy.mock.calls.length).toEqual(1)
  })

  test('write fails due to timeout', async () => {
    const provider = new MockEthereumClient()
    const wallet = new MockEthereumWallet()
    const writeSpy = jest.spyOn(wallet, 'writeContract')
      .mockImplementationOnce(() => {
        throw new Error('Timeout')
      })
    const transactionSpy = jest.spyOn(provider, 'waitForTransactionReceipt')
    const cid = CID.parse('bafyreic5p7grucmzx363ayxgoywb6d4qf5zjxgbqjixpkokbf5jtmdj5ni')
    const tsm = new TransactionStateMachine('test', provider, wallet, CONTRACT, cid)
    const res = await tsm.run()
    expect(res).toMatchSnapshot()
    expect(writeSpy.mock.calls.length).toEqual(2)
    expect(transactionSpy.mock.calls.length).toEqual(1)
  })

  test('transaction receipt fails due to timeout', async () => {
    const provider = new MockEthereumClient()
    const wallet = new MockEthereumWallet()
    const writeSpy = jest.spyOn(wallet, 'writeContract')
    const transactionSpy = jest.spyOn(provider, 'waitForTransactionReceipt')
      .mockImplementationOnce(() => {
        throw new Error('Timeout')
      })
    const blockSpy = jest.spyOn(provider, 'getBlock')
    const cid = CID.parse('bafyreic5p7grucmzx363ayxgoywb6d4qf5zjxgbqjixpkokbf5jtmdj5ni')
    const tsm = new TransactionStateMachine('test', provider, wallet, CONTRACT, cid)
    const res = await tsm.run()
    expect(res).toMatchSnapshot()
    expect(writeSpy.mock.calls.length).toEqual(1)
    expect(transactionSpy.mock.calls.length).toEqual(2)
    expect(blockSpy.mock.calls.length).toEqual(1)
  })

  test('get block fails due to timeout', async () => {
    const provider = new MockEthereumClient()
    const wallet = new MockEthereumWallet()
    const transactionSpy = jest.spyOn(provider, 'waitForTransactionReceipt')
    const blockSpy = jest.spyOn(provider, 'getBlock')
      .mockImplementationOnce(() => {
        throw new Error('Timeout')
      })
    const cid = CID.parse('bafyreic5p7grucmzx363ayxgoywb6d4qf5zjxgbqjixpkokbf5jtmdj5ni')
    const tsm = new TransactionStateMachine('test', provider, wallet, CONTRACT, cid)
    const res = await tsm.run()
    expect(res).toMatchSnapshot()
    expect(transactionSpy.mock.calls.length).toEqual(1)
    expect(blockSpy.mock.calls.length).toEqual(2)
  })

})
