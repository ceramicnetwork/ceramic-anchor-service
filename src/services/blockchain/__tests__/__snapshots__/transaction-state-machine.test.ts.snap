// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionStateMachine with mocks get block fails due to timeout 1`] = `
Transaction {
  "blockNumber": 1n,
  "blockTimestamp": 0n,
  "chain": "test",
  "txHash": "0xdeadbeef",
}
`;

exports[`TransactionStateMachine with mocks no errors encountered 1`] = `
Transaction {
  "blockNumber": 1n,
  "blockTimestamp": 0n,
  "chain": "test",
  "txHash": "0xdeadbeef",
}
`;

exports[`TransactionStateMachine with mocks simulate fails due to timeout 1`] = `
Transaction {
  "blockNumber": 1n,
  "blockTimestamp": 0n,
  "chain": "test",
  "txHash": "0xdeadbeef",
}
`;

exports[`TransactionStateMachine with mocks transaction receipt fails due to timeout 1`] = `
Transaction {
  "blockNumber": 1n,
  "blockTimestamp": 0n,
  "chain": "test",
  "txHash": "0xdeadbeef",
}
`;

exports[`TransactionStateMachine with mocks write fails due to timeout 1`] = `
Transaction {
  "blockNumber": 1n,
  "blockTimestamp": 0n,
  "chain": "test",
  "txHash": "0xdeadbeef",
}
`;
